sdk/containers/azcontainerregistry/blob_custom_client.go

```diff
@@ -94,12 +109,23 @@ type BlobClientUploadChunkOptions struct {
 //   - blobDigestCalculator - Calculator that help to calculate blob digest
 //   - options - BlobClientUploadChunkOptions contains the optional parameters for the BlobClient.UploadChunk method.
 func (client *BlobClient) UploadChunk(ctx context.Context, location string, chunkData io.ReadSeeker, blobDigestCalculator *BlobDigestCalculator, options *BlobClientUploadChunkOptions) (BlobClientUploadChunkResponse, error) {
+	err := blobDigestCalculator.saveState()
+	if err != nil {
+		return BlobClientUploadChunkResponse{}, fmt.Errorf("could not save state of blog digest calculator with error %s", err)
+	}
 	wrappedChunkData := &wrappedReadSeeker{Reader: io.TeeReader(chunkData, blobDigestCalculator.h), Seeker: chunkData}
 	var requestOptions *blobClientUploadChunkOptions
 	if options != nil && options.RangeStart != nil && options.RangeEnd != nil {
 		requestOptions = &blobClientUploadChunkOptions{ContentRange: to.Ptr(fmt.Sprintf("%d-%d", *options.RangeStart, *options.RangeEnd))}
 	}
-	return client.uploadChunk(ctx, location, streaming.NopCloser(wrappedChunkData), requestOptions)
+	resp, err := client.uploadChunk(ctx, location, streaming.NopCloser(wrappedChunkData), requestOptions)
+	if err != nil {
+		restoreErr := blobDigestCalculator.restoreState()
+		if restoreErr != nil {
+			err = fmt.Errorf("could not restore state of blog digest calculator with error %s: %w", restoreErr, err)
```

Comment:
Remove `fmt.Errorf`.

What would happen if `uploadChunk` and `restoreState` fail?  I get that the latter is highly unlikely, I'm just curious.

Reply:
Let me explain. When SDK user call `UploadChunk` several times to upload an image chunk by chunk, only one `BlobDigestCalculator` should be used. So, if `uploadChunk` fail for one of the chunks, the digest calculation of this chunk should be reverted, then user could continue to retry, otherwise the final sum of the image will be wrong. The problem here is restore the hash will also through an error, then we have response error and restore error, I wrapped the response error with restore error and return to user.

Reply:
Thanks for the explanation.

Generally speaking, you should only use `%w` when you want to expose the error as part of the public surface area (i.e. you want to enable `errors.As` and/or `errors.Is` on the wrapped error).  For this error, it's unnecessary as the caller can't programmatically respond to it, but they _can_ respond to the `*azcore.ResponseError` so we should just be returning that (it's the more relevant error).

In addition, the call to `UnmarshalBinary` will only fail if the provided bytes aren't an actual sha256 hash.  Since we have complete control over the provided data, the likelihood of this failing should be impossible.  We still need to surface the error, but not in any special way.

Reply:
I agree. I've refined the code.